### ISSUE TITLE
Failures when running on a 4 GPU server

### ISSUE THREAD

**Issue by guylaporte** [mila-iqia/milabench#183]
When running the nightly of the benchmark on a 4 GPU server, I get these errors:

```
fail   n         perf   sem%   std% peak_memory        score weight
--
bert-fp16                   0   4       274.85   0.0%   0.3%         24919  1099.480993   0.00
bert-fp32                   0   4        83.04   0.3%   2.8%         31867   330.767989   0.00
bert-tf32                   0   4       220.49   0.2%   2.7%         31867   883.275327   0.00
bert-tf32-fp16              0   4       274.94   0.0%   0.2%         24919  1099.772837   3.00
convnext_large-fp16         0     4     529.37   1.6%    17.8%       28247  2147.576559   0.00
convnext_large-fp32         0     4     115.09   2.1%    23.1%       50417   462.315300   0.00
convnext_large-tf32         0     4     239.79   1.0%    10.7%       50417   966.236199   0.00
convnext_large-tf32-fp16    0     4     531.81   1.6%    17.9%       28247  2157.769695   3.00
davit_large                 0   4       538.49   0.2%   3.3%         37145  2155.488634   1.00
davit_large-multi           1     1        NaN    NaN      NaN         563          NaN   5.00
dlrm                        1   1          NaN    NaN    NaN          -1          NaN   1.00
focalnet                    0   4       657.99   0.2%   3.2%         28447  2636.285975   2.00
opt-1_3b                    1   1          NaN    NaN    NaN          -1          NaN   5.00
opt-1_3b-multinode        NaN NaN        NaN      NaN    NaN         NaN          NaN    NaN
opt-6_7b                    1   1          NaN    NaN    NaN          -1          NaN   5.00
opt-6_7b-multinode        NaN NaN        NaN      NaN    NaN         NaN          NaN    NaN
reformer                    0   4        85.81   0.1%   1.2%         25723   343.692581   1.00
regnet_y_128gf              0   4       187.13   0.4%   6.8%         30995   750.787225   2.00
resnet152                   0   4      1172.00   0.3%   4.0%         39697  4695.984761   1.00
resnet152-multi             1   1          NaN    NaN    NaN          -1          NaN   5.00
resnet50                    0   4      1371.35   1.1%  17.4%          5509  5498.089182   1.00
rwkv                        0   4       602.46   0.1%   2.2%          6931  2413.622731   1.00
stargan                     0   4        75.79   1.0%  15.9%         37743   304.917212   1.00
super-slomo                 0   4        76.99   0.4%   5.7%         35421   308.137420   1.00
t5                          0   4        87.08   0.0%   0.3%         35841   348.345862   2.00
whisper                     0   4       387.36   0.1%   2.1%         37737  1550.739555   1.00
 
Scores
------
Failure rate:       6.17% (FAIL)
Score:              31.86
```

If I run version 0.0.7, I get a score of 6 and the same failure rate.  I do not have these errors on a 8 GPU server.  Should the benchmark automatically adjust to the number of GPUs and the total GPU memory?  I will upload the log files.
[logs.tar.gz](https://github.com/mila-iqia/milabench/files/13733269/logs.tar.gz)

**Comment by Delaunay** [comment]
Milabench detects the number of GPU available and run for that number.

I can see errors regarding NCCL in the logs.
It looks like the 4 GPU server might have something misconfigured causing the failures

```
ncclInternalError: Internal check failed.
Last error:
Failed to find reverse path from remNode 0/5f000 nlinks 4 to node 0/cb000
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/milabench/envs/venv/torch/lib/python3.10/site-packages/voir/__main__.py", line 4, in <module>
    main()
  File "/milabench/envs/venv/torch/lib/python3.10/site-packages/voir/cli.py", line 30, in main
    ov(sys.argv[1:] if argv is None else argv)
  File "/milabench/envs/venv/torch/lib/python3.10/site-packages/voir/overseer.py", line 148, in __call__
    super().__call__(*args, **kwargs)
  File "/milabench/envs/venv/torch/lib/python3.10/site-packages/voir/phase.py", line 326, in __call__
    super().__call__(*args, **kwargs)
  File "/milabench/envs/venv/torch/lib/python3.10/site-packages/voir/phase.py", line 268, in __call__
    self.run(*args, **kwargs)
  File "/milabench/envs/venv/torch/lib/python3.10/site-packages/voir/overseer.py", line 143, in run
    set_value(func())
  File "/milabench/envs/venv/torch/lib/python3.10/site-packages/voir/overseer.py", line 195, in <lambda>
    return lambda: exec(mainsection, glb, glb)
  File "/milabench/milabench/benchmarks/timm/pytorch-image-models/train.py", line 1035, in <module>
    main()
  File "/milabench/milabench/benchmarks/timm/pytorch-image-models/train.py", line 562, in main
    model = NativeDDP(model, device_ids=[device], broadcast_buffers=not args.no_ddp_bb)
  File "/milabench/envs/venv/torch/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 674, in __init__
    _verify_param_shape_across_processes(self.process_group, parameters)
  File "/milabench/envs/venv/torch/lib/python3.10/site-packages/torch/distributed/utils.py", line 118, in _verify_param_shape_across_processes
    return dist._verify_params_across_processes(process_group, tensors, logger)
torch.distributed.DistBackendError: NCCL error in: ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1275, internal error, NCCL version 2.14.3
ncclInternalError: Internal check failed.
```

**Comment by guylaporte** [comment]
Here is more information on this issue.  I am now testing version v0.0.8 of the container with CUDA 11.8.  On a 8 GPU server, I don't get any failures (on CUDA 12.3, the rwkv test fails).  On a 4 GPU server with CUDA 11.8, I still get the NCCL errors:

opt-1_3b.0 [stderr] DistBackendError: NCCL error in: ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1275, internal error, NCCL
version 2.14.3
opt-1_3b.0 [stderr] ncclInternalError: Internal check failed.
opt-1_3b.0 [stderr] Last error:
opt-1_3b.0 [stderr] Failed to find reverse path from remNode 0/5f000 nlinks 4 to node 0/cb000

As a test, on the same exact 4 GPU server and configuration, I installed the benchmark from the master branch and ran the test without using a container.  All tests passed.  The main difference is that this method uses the NCCL version 2.18.5+cuda11.8 as opposed to version 2.14.3 used by the container.  Would it be possible to build the container using NCCL version 2.18.5+cuda11.8 please?

**Comment by guylaporte** [comment]
The release notes for NCCL 2.15.1 which shows when official H100 support was added:

https://docs.nvidia.com/deeplearning/nccl/release-notes/rel_2-15-1.html#rel_2-15-1

**Comment by Delaunay** [comment]
Fixed in https://github.com/mila-iqia/milabench/pull/191 released in v0.0.9
